### PR TITLE
[models] Unleash FAST model (reparameterization logic)

### PR DESCRIPTION
**Pull request by felixdittrich92**
This PR:

- Adds inference only reparameterization logic
- Implements the optimization mentioned in the paper
- Not activated by default atm

Note: reparameterization is provided as function instead of a flag while model initialization the reason for this is that this would not work with custom loaded user model so i think that's actually the better way :)

Any feedback is welcome :hugs:

### CODE REVIEW

File: docs/source/using_doctr/using_models.rst [hunk 1]
```diff
@@ -34,45 +34,45 @@ The following architectures are currently supported:
...
++----------------+---------------------------------+-----------------+---------------+------------+---------------+------------+---------------+--------------------+
+| TensorFlow     | fast_tiny                       | (1024, 1024, 3) | 13.5 M (8.5M) |            |               |            |               | 0.7 (0.4)          |
++----------------+---------------------------------+-----------------+---------------+------------+---------------+------------+---------------+--------------------+
+| TensorFlow     | fast_small                      | (1024, 1024, 3) | 14.7 M (9.7M) |            |               |            |               | 0.7 (0.5)          |
++----------------+---------------------------------+-----------------+---------------+------------+---------------+------------+---------------+--------------------+
+| TensorFlow     | fast_base                       | (1024, 1024, 3) | 16.3 M (10.6M)|            |               |            |               | 0.8 (0.5)          |
++----------------+---------------------------------+-----------------+---------------+------------+---------------+------------+---------------+--------------------+
```
Comment (odulcy-mindee) [inline docs/source/using_doctr/using_models.rst:57]: What are the values in parenthesis ?

File: doctr/models/detection/fast/pytorch.py [hunk 2]
```diff
@@ -279,6 +279,47 @@ def ohem_sample(score: torch.Tensor, gt: torch.Tensor, mask: torch.Tensor) -> to
         return text_loss + kernel_loss
 
 
+def reparameterize(model: Union[FAST, nn.Module]) -> FAST:
```
Comment (odulcy-mindee) [inline doctr/models/detection/fast/pytorch.py:282]: I don't understand what is the purpose of reparameterization.
First, we need to instantiate the model, then to modify some layers. Is it for inference purpose ?

File: doctr/models/modules/layers/pytorch.py [hunk 3]
```diff
@@ -84,3 +90,77 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         id_out = self.rbr_identity(x) if self.rbr_identity is not None and self.ver_bn is not None else 0
 
         return self.activation(main_outputs + vertical_outputs + horizontal_outputs + id_out)
+
+    # The following logic is used to reparmetrize the layer
```
Comment (odulcy-mindee) [inline doctr/models/modules/layers/pytorch.py:94]: typo

File: doctr/models/modules/layers/tensorflow.py [hunk 4]
```diff
@@ -93,3 +100,76 @@ def call(self, x: tf.Tensor, **kwargs: Any) -> tf.Tensor:
         id_out = self.rbr_identity(x, **kwargs) if self.rbr_identity is not None and self.ver_bn is not None else 0
 
         return self.activation(main_outputs + vertical_outputs + horizontal_outputs + id_out)
+
+    # The following logic is used to reparmetrize the layer
```
Comment (odulcy-mindee) [inline doctr/models/modules/layers/tensorflow.py:104]: typo

File: doctr/models/modules/layers/pytorch.py [hunk 5]
```diff
@@ -84,3 +90,77 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        kernel_mxn = kernel_mxn + kernel_mx1 + kernel_1xn + kernel_id
+        bias_mxn = bias_mxn + bias_mx1 + bias_1xn + bias_id
+        return kernel_mxn, bias_mxn
+
+    def _pad_to_mxn_tensor(self, kernel: torch.Tensor) -> torch.Tensor:
+        kernel_height, kernel_width = self.converted_ks
+        height, width = kernel.shape[2:]
+        pad_left_right = (kernel_width - width) // 2
+        pad_top_down = (kernel_height - height) // 2
+        return torch.nn.functional.pad(kernel, [pad_left_right, pad_left_right, pad_top_down, pad_top_down])
```
Comment (odulcy-mindee) [inline doctr/models/modules/layers/pytorch.py:140]: You can explicitly set `value=0` for padding so we're sure it's the same value than the one in TF

File: doctr/models/modules/layers/tensorflow.py [hunk 6]
```diff
@@ -93,3 +100,76 @@ def call(self, x: tf.Tensor, **kwargs: Any) -> tf.Tensor:
...
+        kernel = self.id_tensor
+        std = tf.sqrt(identity.moving_variance + identity.epsilon)
+        t = tf.reshape(identity.gamma / std, (-1, 1, 1, 1))
+        return kernel * t, identity.beta - identity.moving_mean * identity.gamma / std
+
+    def _fuse_bn_tensor(self, conv: layers.Conv2D, bn: layers.BatchNormalization) -> Tuple[tf.Tensor, tf.Tensor]:
+        kernel = conv.kernel
+        kernel = self._pad_to_mxn_tensor(kernel)
+        std = tf.sqrt(bn.moving_variance + bn.epsilon)
+        t = tf.reshape(bn.gamma / std, (1, 1, 1, -1))
```
Comment (odulcy-mindee) [inline doctr/models/modules/layers/tensorflow.py:127]: Why is it `(1, 1, 1, -1)` instead of `(-1, 1, 1, 1)` ?

**Review by odulcy-mindee — CHANGES_REQUESTED** [review]

File: doctr/models/detection/fast/pytorch.py (hunk 2 above)
Comment (felixdittrich92) [reply doctr/models/detection/fast/pytorch.py:282]: yep in simple words it replaces all FASTConvLayers with simplier Conv2D layers, merges the weights and set init the Conv2D with it 

reparameterization is only onces applied before inference :)

File: docs/source/using_doctr/using_models.rst (hunk 1 above)
Comment (felixdittrich92) [reply docs/source/using_doctr/using_models.rst:57]: The values from the reparameterized model 

It's smaller and much faster by using the inference only reparameterized version

File: doctr/models/modules/layers/tensorflow.py (hunk 6 above)
Comment (felixdittrich92) [reply doctr/models/modules/layers/tensorflow.py:127]: the shapes in TF and PT are different ^^

TF: (kernel_size, kernel_size, dim, filter_dim)
PT: (filter_dim, dim, kernel_size, kernel_size)

**Review by odulcy-mindee — APPROVED** [review]
